### PR TITLE
Request header invalid 6415 v3

### PR DESCRIPTION
**Pull request by catenacyber**
https://redmine.openinfosecfoundation.org/issues/6415

#409 with unit tests to new tokenization function

### CODE REVIEW

File: htp/htp_util.c [hunk 1]
```diff
@@ -2536,3 +2536,54 @@ htp_uri_t *htp_uri_alloc(void) {
...
+                    // start of next token
+                    state = 0;
+                }
+                break;
+            case 2:
+                if (hvp[i] == ',') {
+                    return HTP_OK;
+                }
+                if (hvp[i] != 0 && hvp[i] != ' ') {
+                    // trailing junk in token, cait for a next one
```
Comment (victorjulien) [inline htp/htp_util.c:2579]: typo: cait should be wait?

File: htp/htp_util.c [hunk 2]
```diff
@@ -2536,3 +2536,54 @@ htp_uri_t *htp_uri_alloc(void) {
 char *htp_get_version(void) {
     return HTP_VERSION_STRING_FULL;
 }
+
+htp_status_t htp_header_has_token(const unsigned char *hvp, size_t hvlen, const unsigned char *value) {
```
Comment (victorjulien) [inline htp/htp_util.c:2540]: please provide documentation of the function

Comment (victorjulien) [inline htp/htp_util.c:2540]: hvp is haystack, value is needle?

File: htp/htp_util.c [hunk 3]
```diff
@@ -2536,3 +2536,54 @@ htp_uri_t *htp_uri_alloc(void) {
...
+    // offset to compare in value
+    size_t v_off = 0;
+    for (size_t i = 0; i < hvlen; i++) {
+        switch (state) {
+            case 0:
+                if (hvp[i] == 0) {
+                    // skip 0 char
+                    continue;
+                }
+                if (v_off == 0 && hvp[i] == ' ') {
```
Comment (victorjulien) [inline htp/htp_util.c:2551]: should we use htp_is_space() or similar to be more liberal about spaces?

File: htp/htp_util.c [hunk 4]
```diff
@@ -2536,3 +2536,54 @@ htp_uri_t *htp_uri_alloc(void) {
...
+                    // skip leading space
+                    continue;
+                }
+                if (tolower(hvp[i]) == tolower(value[v_off])) {
```
Comment (victorjulien) [inline htp/htp_util.c:2555]: func should document that it does a caseless search

also, if needle is always passed as lowercase we can skip the tolower here

**Review by victorjulien — CHANGES_REQUESTED** [review]

File: htp/htp_util.c (hunk 1 above)
Comment (catenacyber) [reply htp/htp_util.c:2579]: indeed

File: htp/htp_util.c (hunk 2 above)
Comment (catenacyber) [reply htp/htp_util.c:2540]: ok

Comment (catenacyber) [reply htp/htp_util.c:2540]: yes, and haystack is header value, value is token to look for

File: htp/htp_util.c (hunk 3 above)
Comment (catenacyber) [reply htp/htp_util.c:2551]: Looks a good idea

File: htp/htp_util.c (hunk 4 above)
Comment (catenacyber) [reply htp/htp_util.c:2555]: indeed

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/libhtp/pull/413

**Pull request closed without merging**